### ISSUE TITLE
New Rule: Use Type Checking instead of Type Casting

### ISSUE THREAD

**Issue by schalkms** [detekt/detekt#7860]
## Expected Behavior of the rule

The explicit type cast can be removed to optimise the check.
The compliant case checks for the type only, while the noncompliant case first checks if the instance is of that type and then returns null or casts explicitly to that type.
The rule should not flag cases, where the result of the cast operator, e.g. the casted type variable, is used in subsequent code.

#### Compliant

```Kotlin
val objList: List<Any> = emptyList()
objList.any { it is String }
```

#### Noncompliant

```Kotlin
val objList: List<Any> = emptyList()
objList.any { it as? String != null }
```

They might be more complicated cases, where this rule could be helpful.

## Context

The `performance` rule set should include this new rule.
The idea was initiated in #7858.

**Comment by BraisGabin** [comment]
This could go too to style. The compliant code is easier to read.

**Comment by schalkms** [comment]
> This could go too to style. The compliant code is easier to read.

I don't think so. The noncompliant code compiles to more byte code than necessary, because not only a type is performed but also a cast. If such a cast happens in a long list, this impacts run time and memory consumption.
